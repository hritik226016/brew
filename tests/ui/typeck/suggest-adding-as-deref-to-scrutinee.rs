use std::io::ErrorKind;

type E = Vec<ErrorKind>;

fn res(res: Result<Vec<i32>, E>) {
    match res {
        //~^ HELP: consider using `as_deref` here
        Ok([1]) => true,
        //~^ ERROR: expected an array or slice
        Err(_) => false,
    };

    match res {
        //~^ HELP: consider using `as_deref` here
        Ok([1]) => true,
        //~^ ERROR: expected an array or slice
        Err([ErrorKind::NotFound]) => false,
        //~^ ERROR: expected an array or slice
    };
}

fn opt(opt: Option<Vec<i32>>) {
    match opt {
        //~^ HELP: consider using `as_deref` here
        Some([1]) => true,
        //~^ ERROR: expected an array or slice
        None => false,
    };
}

fn err_arm_only(res: Result<Vec<i32>, E>) {
    // doesn't suggest `as_deref` since the arm is `Err`,
    // which `as_deref` doesn't make sense
    match res {
        Err([ErrorKind::NotFound]) => true,
        //~^ ERROR: expected an array or slice
        _ => false,
    };
}

fn res_cannot_be_array_or_slice(res: Result<i32, E>) {
    // doesn't suggest `as_deref` since `T` in `Ok(T)` can't be dereferenced
    // to an array or slice
    match res {
        Ok([1]) => true,
        //~^ ERROR: expected an array or slice
        _ => false,
    };
}

fn opt_cannot_be_array_or_slice(opt: Option<i32>) {
    // doesn't suggest `as_deref` since `T` in `Some(T)` can't be dereferenced
    // to an array or slice
    match opt {
        Some([1]) => true,
        //~^ ERROR: expected an array or slice
        None => false,
    };
}

fn res_has_ref(res: Result<&Vec<i32>, E>) {
    // `T` in `Ok(T)` can't be dereferenced to an array or slice
    // because `res.as_deref()` would return `Result<&Vec<i32>, &E>`
    match res {
        Ok([1]) => true,
        //~^ ERROR: expected an array or slice
        Err(_) => false,
    };
}

fn opt_has_ref(opt: Option<&Vec<i32>>) {
    // `T` in `Some(T)` can't be dereferenced to an array or slice
    // because `opt.as_deref()` would return `Option<&Vec<i32>>`
    match opt {
        Some([1]) => true,
        //~^ ERROR: expected an array or slice
        None => false,
    };
}

fn main() {}
