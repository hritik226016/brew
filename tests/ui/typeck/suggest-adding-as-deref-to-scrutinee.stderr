error[E0529]: expected an array or slice, found `Vec<i32>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:8:12
   |
LL |         Ok([1]) => true,
   |            ^^^ pattern cannot match with input type `Vec<i32>`
   |
help: consider using `as_deref` here
   |
LL |     match res.as_deref() {
   |              +++++++++++

error[E0529]: expected an array or slice, found `Vec<i32>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:15:12
   |
LL |         Ok([1]) => true,
   |            ^^^ pattern cannot match with input type `Vec<i32>`
   |
help: consider using `as_deref` here
   |
LL |     match res.as_deref() {
   |              +++++++++++

error[E0529]: expected an array or slice, found `Vec<ErrorKind>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:17:13
   |
LL |         Err([ErrorKind::NotFound]) => false,
   |             ^^^^^^^^^^^^^^^^^^^^^ pattern cannot match with input type `Vec<ErrorKind>`

error[E0529]: expected an array or slice, found `Vec<i32>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:25:14
   |
LL |         Some([1]) => true,
   |              ^^^ pattern cannot match with input type `Vec<i32>`
   |
help: consider using `as_deref` here
   |
LL |     match opt.as_deref() {
   |              +++++++++++

error[E0529]: expected an array or slice, found `Vec<ErrorKind>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:35:13
   |
LL |         Err([ErrorKind::NotFound]) => true,
   |             ^^^^^^^^^^^^^^^^^^^^^ pattern cannot match with input type `Vec<ErrorKind>`

error[E0529]: expected an array or slice, found `i32`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:45:12
   |
LL |         Ok([1]) => true,
   |            ^^^ pattern cannot match with input type `i32`

error[E0529]: expected an array or slice, found `i32`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:55:14
   |
LL |         Some([1]) => true,
   |              ^^^ pattern cannot match with input type `i32`

error[E0529]: expected an array or slice, found `Vec<i32>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:65:12
   |
LL |         Ok([1]) => true,
   |            ^^^ pattern cannot match with input type `Vec<i32>`

error[E0529]: expected an array or slice, found `Vec<i32>`
  --> $DIR/suggest-adding-as-deref-to-scrutinee.rs:75:14
   |
LL |         Some([1]) => true,
   |              ^^^ pattern cannot match with input type `Vec<i32>`

error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0529`.
